analytics_export/analyticsExport.js

```diff
@@ -1,8 +1,197 @@
+const url = require('url');
+const util = require('util');
+
+const { BigqueryClient } = require('./bigqueryClient');
+const { measureToTablePrefix, dimensionToTableSuffix } = require('./tableMetadata');
+const { RequestError } = require('./requestError');
+
+class AnalyticsExport {
+  constructor(client, project, dataset, appId, appName) {
+    this.client = client;
+    this.project = project;
+    this.dataset = dataset;
+    this.appId = appId;
+    this.appName = appName;
+  }
+
+  /**
+   * Return a Map where key is a dimension key and value is an array of measure keys
+   * that can be grouped by the dimension
+   */
+  async getAllowedDimensionsPerMeasure(startTimestamp, endTimestamp) {
+    const getSettings = util.promisify(this.client.getSettings).bind(this.client);
+    const settings = await getSettings();
+
+    const dataEndDate = settings.configuration.dataEndDate.slice(0, 10);
+    const dataStartDate = settings.configuration.dataStartDate.slice(0, 10);
+
+    if (endTimestamp > Date.parse(dataEndDate) || startTimestamp < Date.parse(dataStartDate)) {
+      throw Error(`Date out of range; data exists for ${dataStartDate} to ${dataEndDate}`);
+    }
+
+    const measuresByDimension = new Map();
+    const dimensionsById = {};
+
+    // Create number id to key mapping
+    settings.dimensions.forEach((dimension) => {
+      if (dimension.groupBy) {
+        dimensionsById[dimension.id] = dimension.key;
+        measuresByDimension.set(dimension.key, []);
+      }
+    });
+
+    // Set null dimension to get ungrouped totals
+    measuresByDimension.set(null, []);
+
+    // Create mapping of dimensions to allowed measures
+    settings.measures.forEach((measure) => {
+      measure.dimensions.forEach((dimensionId) => {
+        if (dimensionId in dimensionsById) {
+          measuresByDimension.get(dimensionsById[dimensionId])
+            .push(measure.key);
+        }
+      });
+      measuresByDimension.get(null).push(measure.key);
+    });
+
+    return measuresByDimension;
+  }
+
+  async getMetric(startDate, endDate, measure, dimension) {
+    console.log(`Getting ${measure} by ${dimension}`);
+
+    const dimensionGiven = dimension !== undefined && dimension !== null;
+
+    const measures = measure instanceof Array ? measure : [measure];
+    const queryConfig = {
+      measures,
+      frequency: itc.frequency.days,
+    };
+    if (dimensionGiven) {
+      queryConfig.group = { dimension };
+    }
+    const query = itc.AnalyticsQuery.metrics(this.appId, queryConfig).date(startDate, endDate);
+
+    // Directly fetch instead of itc.request to avoid queue and properly handle errors
+    const response = await fetch(
+      url.parse(query.apiURL + query.endpoint),
+      {
+        method: 'POST',
+        body: JSON.stringify(query.assembleBody()),
+        headers: this.client.getHeaders(),
+      },
+    );
+    const data = await response.json();
+    if (response.ok === false) {
```

Comment:
I think this less verbose form should still work

```suggestion
    if (!response.ok) {
```